ehcache-impl/src/main/java/org/ehcache/impl/internal/loaderwriter/writebehind/BatchingLocalHeapWriteBehindQueue.java

```diff
@@ -188,13 +193,19 @@ public boolean add(SingleOperation<K, V> operation) {
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public void run() {
       try {
         List<BatchOperation<K, V>> batches = createMonomorphicBatches(operations());
         // execute the batch operations
         for (BatchOperation<K, V> batch : batches) {
           try {
             batch.performOperation(cacheLoaderWriter);
+          }  catch (BulkCacheWritingException e) {
+            Iterator<K> it = (Iterator<K>) e.getFailures().keySet().iterator();
+            while(it.hasNext()) {
+              keyCleanUpMethod.accept(it.next());
+            }
           } catch (Exception e) {
```

Comment:
You need to cleanup on the regular exception path too. It's legal for a writer to throw any exception and expect cleanup to be done.